osu.Game/Graphics/ParticleSpewer.cs

```diff
@@ -0,0 +1,208 @@
+using osu.Framework.Graphics.OpenGL.Vertices;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.Textures;
+using osuTK;
+
+namespace osu.Game.Graphics
+{
+    public class ParticleSpewer : Sprite
+    {
+        private readonly FallingParticle[] particles;
+        private int currentIndex;
+        private double lastParticleAdded;
+
+        private readonly double cooldown;
+        private readonly double maxDuration;
+
+        /// <summary>
+        /// Determines whether particles are being spawned.
+        /// </summary>
+        public readonly BindableBool Active = new BindableBool();
+
+        public override bool IsPresent => base.IsPresent && hasActiveParticles;
+
+        /// <summary>
+        /// Called each time a new particle should be spawned.
+        /// </summary>
+        public Func<FallingParticle?> CreateParticle = () => new FallingParticle();
+
+        public float ParticleGravity;
+
+        private bool hasActiveParticles => Active.Value || (lastParticleAdded + maxDuration) > Time.Current;
+
+        public ParticleSpewer(Texture texture, int perSecond, double maxDuration)
+        {
+            Texture = texture;
+            Blending = BlendingParameters.Additive;
+
+            particles = new FallingParticle[perSecond * (int)Math.Ceiling(maxDuration / 1000)];
+
+            cooldown = 1000f / perSecond;
+            this.maxDuration = maxDuration;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            // reset cooldown if the clock was rewound.
+            // this can happen when seeking in replays.
+            if (lastParticleAdded > Time.Current) lastParticleAdded = 0;
+
+            if (Active.Value && Time.Current > lastParticleAdded + cooldown)
+            {
+                var newParticle = CreateParticle();
+
+                if (newParticle.HasValue)
+                {
+                    var particle = newParticle.Value;
+                    particle.StartTime = (float)Time.Current;
+
+                    particles[currentIndex] = particle;
+
+                    currentIndex = (currentIndex + 1) % particles.Length;
+                    lastParticleAdded = Time.Current;
+                }
+            }
+
+            Invalidate(Invalidation.DrawNode);
+        }
+
+        protected override DrawNode CreateDrawNode() => new ParticleSpewerDrawNode(this);
+
+        # region DrawNode
+
+        private class ParticleSpewerDrawNode : SpriteDrawNode
+        {
+            private readonly FallingParticle[] particles;
+
+            protected new ParticleSpewer Source => (ParticleSpewer)base.Source;
+
+            private readonly float maxDuration;
+
+            private float currentTime;
+            private float gravity;
+            private Axes relativePositionAxes;
+            private Vector2 sourceSize;
+
+            public ParticleSpewerDrawNode(Sprite source)
```

Comment:
Not sure if this is just me but with the cast in `Source` (the property) I'd prefer that this ctor parameter was scoped down to `ParticleSpewer`. I was really confused for a second as to what a sprite was doing here.